### PR TITLE
SharedArrayBuffer - Initial work.

### PR DESCRIPTION
**Pull request by akroshg**
This work is to implement a prototype version of the `SharedArrayBuffer`. The spec is in the stage2.
The `SharedArrayBuffer` is behind the `ESSharedArrayBuffer` (or -sab)
flag.
Highlights.
Introduce the `SharedArrayBuffer` type and its implementation.
Refactor the `ArrayBuffer` to a common class (`ArrayBufferBase`) so that both `SharedArrayBuffer` and `ArrayBuffer` leverage the common functionality and machinery.
`Atomics` object (spec'ed) is introduced to provide the atomic operation on the buffer
which is shared. Currently it is using the Win32 based APIs to provide the
functionality.
All 12 methods of `Atomics` are implemented.
All `TypedArray` views are changed to make use of `SharedArrayBuffer` as well.
The `Serialization/Deserialization` implementation is in the different
repo.
Added test cases to validate most of the functionality. sharedarraybuffer - initial work

### CODE REVIEW

**Comment by akroshg** [comment]
Spec : http://tc39.github.io/ecmascript_sharedmem/shmem.html

**Comment by akroshg** [comment]
I'll take a look at the Ubuntu failures. I haven't made any effort to have this x-plat in this V1 implementation but it should just clean build everywhere.

File: lib/Common/ConfigFlagsList.h [hunk 1]
```diff
@@ -783,6 +784,14 @@ PHASE(All)
 #define FLAGPR(Type, ParentName, Name, String, Default)     FLAG(Type, Name, String, Default, ParentName, FALSE)
 #define FLAGR(Type, Name, String, Default)                  FLAG(Type, Name, String, Default, NoParent, FALSE)
 
+// Release flags with paraent and acronym
```
Comment (agarwal-sandeep) [inline lib/Common/ConfigFlagsList.h]: >paraent [](start = 22, length = 7)

parent

File: lib/Parser/rterrors.h [hunk 2]
```diff
@@ -357,3 +357,10 @@ RT_ERROR_MSG(JSERR_InvalidHint, 5658, "%s: invalid hint", "invalid hint", kjstTy
 
 RT_ERROR_MSG(JSERR_This_NeedNamespace, 5659, "%s: 'this' is not a Module Namespace object", "Module Namespace object expected", kjstTypeError, JSERR_This_NeedNamespace) // {Locked="\'this\'"}
 RT_ERROR_MSG(JSERR_This_NeedListIterator, 5660, "%s: 'this' is not a List Iterator object", "List Iterator expected", kjstTypeError, 0) 
+RT_ERROR_MSG(JSERR_NeedSharedArrayBufferObject, 5661, "%s is not a SharedArrayBuffer", "SharedArrayBuffer object expected", kjstTypeError, 0)
+
+RT_ERROR_MSG(JSERR_Function_LessArguments, 5662, "Function '%s' is called with less arguments", "Function called with less arguments", kjstRangeError, 0)
+RT_ERROR_MSG(JSERR_NeedTypedArrayObject, 5663, "", "Atomics function called with invalid typed array object", kjstTypeError, 0)
+RT_ERROR_MSG(JSERR_InvalidTypedArrayIndex, 5664, "", "access index is out of range", kjstRangeError, 0)
```
Comment (agarwal-sandeep) [inline lib/Parser/rterrors.h]: >JSERR_InvalidTypedArrayIndex [](start = 13, length = 28)

We already have JSERR_ArgumentOutOfRange which can be used instead of creating a new error

File: lib/Common/ConfigFlagsList.h [hunk 3]
```diff
@@ -1012,7 +1021,8 @@ FLAGPR           (Boolean, ES6, ES6Verbose             , "Enable ES6 verbose tra
 #endif
 FLAGPR_REGOVR_EXP(Boolean, ES6, ArrayBufferTransfer    , "Enable ArrayBuffer.transfer"                              , DEFAULT_CONFIG_ArrayBufferTransfer)
 
-FLAGPR           (Boolean, ES6, ESObjectGetOwnPropertyDescriptors, "Enable Object.getOwnPropertyDescriptors"              , DEFAULT_CONFIG_ESObjectGetOwnPropertyDescriptors)
+FLAGPR           (Boolean, ES6, ESObjectGetOwnPropertyDescriptors, "Enable Object.getOwnPropertyDescriptors"        , DEFAULT_CONFIG_ESObjectGetOwnPropertyDescriptors)
+FLAGPRA          (Boolean, ES6, ESSharedArrayBuffer    , sab     , "Enable SharedArrayBuffer"                       , DEFAULT_CONFIG_ESSharedArrayBuffer)
```
Comment (agarwal-sandeep) [inline lib/Common/ConfigFlagsList.h]: >FLAGPRA [](start = 0, length = 7)

Will we be able to disable this using COMPILE_DISABLE_ macro similar to other release ES6 options above?

File: lib/Runtime/Runtime.h [hunk 4]
```diff
@@ -496,7 +496,9 @@ enum tagDEBUG_EVENT_INFO_TYPE
 #include "Library/JavascriptError.h"
 #include "Library/JavascriptArray.h"
 
+#include "Library\AtomicsObject.h"
```
Comment (agarwal-sandeep) [inline lib/Runtime/Runtime.h]: >\ [](start = 17, length = 1)

forward slash for consistency

Comment (jianchun) [inline lib/Runtime/Runtime.h]: good catch, and `\` fails on Linux

File: lib/Runtime/Library/AtomicsObject.cpp [hunk 5]
```diff
@@ -0,0 +1,270 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#include "RuntimeLibraryPch.h"
+
+#define ATOMICS_FUNCTION_BEGIN_WRAPPER(length, methodName) \
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/AtomicsObject.cpp]: >ATOMICS_FUNCTION_BEGIN_WRAPPER [](start = 8, length = 30)

Consider renaming to ATOMICS_FUNCTION_ENTRY_CHECKS

File: lib/Runtime/Library/AtomicsObject.cpp [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+            double dblValue = JavascriptConversion::ToNumber(requestIndex, scriptContext);
+            if (dblValue != accessIndex)
+            {
+                JavascriptError::ThrowRangeError(scriptContext, JSERR_InvalidTypedArrayIndex);
+            }
+        }
+
+        Assert(TypedArrayBase::Is(typedArray));
+        
+        if (accessIndex < 0 || accessIndex >= (int32)TypedArrayBase::FromVar(typedArray)->GetLength())
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/AtomicsObject.cpp]: >GetLength [](start = 90, length = 9)

Can ToInt32_Full/ToNumber call user script ? if so please make sure GetLength returns 0 if the buffer is detached by calls to ToInt32_Full/ToNumber

File: lib/Runtime/Library/AtomicsObject.cpp [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+    TypedArrayBase * AtomicsObject::ValidateAndGetTypedArray(Var typedArray, Var index, __out uint32 *accessIndex, ScriptContext *scriptContext, bool onlyInt32)
+    {
+        ValidateSharedIntegerTypedArray(typedArray, scriptContext, onlyInt32);
+        uint32 i = ValidateAtomicAccess(typedArray, index, scriptContext);
+        if (accessIndex != nullptr)
+        {
+            *accessIndex = i;
+        }
+        Assert(TypedArrayBase::Is(typedArray));
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/AtomicsObject.cpp]: > Assert(TypedArrayBase::Is(typedArray)); [](start = 7, length = 40)

TypedArrayBase::FromVar already have AssertMsg(Is(aValue), "must be a typed array"); no need for this, remove from all places

File: lib/Runtime/Library/AtomicsObject.cpp [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(4, "Atomics.compareExchange");
+
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext);
+
+        Var comparand = args[3];
+        Var replacementValue = args[4];
+
+        return typedArrayBase->TypedCompareExchange(accessIndex, comparand, replacementValue);
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/AtomicsObject.cpp]: >return typedArrayBase->TypedCompareExchange(accessIndex, comparand, replacementValue); [](start = 8, length = 86)

Why create local variables for args[3] and args[4], pass directly (as done elsewhere in this file) for consistency.

File: lib/Runtime/Library/AtomicsObject.cpp [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext);
+        return typedArrayBase->TypedExchange(accessIndex, args[3]);
+    }
+
+    Var AtomicsObject::EntryIsLockFree(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(1, "Atomics.isLockFree");
+        uint32 size = JavascriptConversion::ToUInt32(args[1], scriptContext);
+        return (size == 1 || size == 2 || size == 4) ? scriptContext->GetLibrary()->GetTrue() : scriptContext->GetLibrary()->GetFalse();
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/AtomicsObject.cpp]: >size == 1 || size == 2 || size == 4 [](start = 16, length = 35)

Can we put a comment here explaining why check for 1,2 and 4

File: lib/Runtime/Library/AtomicsObject.cpp [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+
+    Var AtomicsObject::EntryWait(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(3, "Atomics.wait");
+
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext, true /*onlyInt32*/);
+
+        Var value = args[3];
+        Var timeoutVar = args[4];
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/AtomicsObject.cpp]: >Var timeoutVar = args[4]; [](start = 8, length = 25)

Accessing index without length check? above wrapper have length check for 3. Even if this is optional shouldn't we do a callInfo count check?

File: lib/Runtime/Library/AtomicsObject.cpp [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+        uint32 timeout = INFINITE;
+
+        if (args.Info.Count > 4 && !JavascriptOperators::IsUndefinedObject(timeoutVar))
+        {
+            double t =JavascriptConversion::ToNumber(timeoutVar, scriptContext);
+            if (!(NumberUtilities::IsNan(t) || JavascriptNumber::IsPosInf(t)))
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/AtomicsObject.cpp]: >NumberUtilities::IsNan(t) || JavascriptNumber::IsPosInf(t) [](start = 18, length = 58)

Does spec have special handling for NaN and PosInf? what about NegInf?

File: lib/Runtime/Library/AtomicsObject.cpp [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+            }
+        }
+
+        if (!AgentOfBuffer::AgentCanSuspend(scriptContext))
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_CannotSuspendBuffer);
+        }
+
+        Assert(typedArrayBase->GetBytesPerElement() == 4);
+        uint32 bufferIndex = (accessIndex * 4) + typedArrayBase->GetByteOffset();
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/AtomicsObject.cpp]: Probably at this point we need a detach/length check as above we have ToNumber/ToInt32 calls which I think can call user code?

File: lib/Runtime/Library/AtomicsObject.cpp [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+        Assert(bufferIndex < typedArrayBase->GetArrayBuffer()->GetByteLength());
+        SharedArrayBuffer *sharedArrayBuffer = typedArrayBase->GetArrayBuffer()->GetAsSharedArrayBuffer();
+        WaiterList *waiterList = sharedArrayBuffer->GetWaiterList(bufferIndex);
+
+        int32 v = JavascriptConversion::ToInt32(value, scriptContext);
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/AtomicsObject.cpp]: >int32 v = JavascriptConversion::ToInt32(value, scriptContext); [](start = 8, length = 62)

Move these before timeoutVar check as spec have this (step 3) before the timeout check (step 4) and ToInt32 can potentially throw

File: lib/Runtime/Library/Chakra.Runtime.Library.vcxproj [hunk 14]
```diff
@@ -157,11 +157,14 @@
     <ClCompile Include="$(MSBuildThisFileDirectory)IntlEngineInterfaceExtensionObject.cpp" />
     <ClCompile Include="$(MSBuildThisFileDirectory)JavascriptErrorDebug.cpp" />
     <ClCompile Include="$(MSBuildThisFileDirectory)JavascriptListIterator.cpp" />
+    <ClCompile Include="AtomicsObject.cpp" />
+    <ClCompile Include="SharedArrayBuffer.cpp" />
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/Chakra.Runtime.Library.vcxproj]: Need to have $(MSBuildThisFileDirectory)

File: lib/Runtime/Library/Chakra.Runtime.Library.vcxproj.filters [hunk 15]
```diff
@@ -112,6 +112,8 @@
     <ClCompile Include="$(MSBuildThisFileDirectory)JavascriptSimdBool8x16.cpp" />
     <ClCompile Include="$(MSBuildThisFileDirectory)SimdBool8x16Lib.cpp" />
     <ClCompile Include="$(MSBuildThisFileDirectory)JavascriptListIterator.cpp" />
+    <ClCompile Include="SharedArrayBuffer.cpp" />
+    <ClCompile Include="AtomicsObject.cpp" />
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/Chakra.Runtime.Library.vcxproj.filters]: Need $(MSBuildThisFileDirectory)

File: lib/Parser/rterrors.h (hunk 2 above)
Comment (akroshg) [inline lib/Parser/rterrors.h]: I like more descriptive errors messages. We are actually accessing the index of the buffer - hence the error.

---
In reply to: [77394931](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77394931) [](ancestors = 77394931)

File: lib/Common/ConfigFlagsList.h (hunk 3 above)
Comment (akroshg) [inline lib/Common/ConfigFlagsList.h]: I believe this is not even as experimental phase. None of ES7(ESNext) are having the COMPILE_DISABLE macro. So I put that as is.

---
In reply to: [77396110](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77396110) [](ancestors = 77396110)

File: lib/Runtime/Runtime.h (hunk 4 above)
Comment (akroshg) [inline lib/Runtime/Runtime.h]: good catch. thanks

---
In reply to: [77396220](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77396220) [](ancestors = 77396220)

File: lib/Runtime/Library/AtomicsObject.cpp (hunk 5 above)
Comment (akroshg) [inline lib/Runtime/Library/AtomicsObject.cpp]: agreed.

---
In reply to: [77402371](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77402371) [](ancestors = 77402371)

File: lib/Runtime/Library/AtomicsObject.cpp (hunk 6 above)
Comment (akroshg) [inline lib/Runtime/Library/AtomicsObject.cpp]: I have test cases for that. Will you check that and verify if that is what you meant?

---
In reply to: [77403224](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77403224) [](ancestors = 77403224)

File: lib/Runtime/Library/AtomicsObject.cpp (hunk 7 above)
Comment (akroshg) [inline lib/Runtime/Library/AtomicsObject.cpp]: will do.

---
In reply to: [77404054](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77404054) [](ancestors = 77404054)

File: lib/Runtime/Library/AtomicsObject.cpp (hunk 9 above)
Comment (akroshg) [inline lib/Runtime/Library/AtomicsObject.cpp]: It is somewhat explained in the spec - but will add comment over here.

---
In reply to: [77404558](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77404558) [](ancestors = 77404558)

File: lib/Runtime/Library/AtomicsObject.cpp (hunk 8 above)
Comment (akroshg) [inline lib/Runtime/Library/AtomicsObject.cpp]: will do. thanks,

---
In reply to: [77404429](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77404429) [](ancestors = 77404429)

File: lib/Runtime/Library/AtomicsObject.cpp (hunk 10 above)
Comment (akroshg) [inline lib/Runtime/Library/AtomicsObject.cpp]: ah - good point. I have check below. I don't need this variable.

---
In reply to: [77404853](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77404853) [](ancestors = 77404853)

File: lib/Runtime/Library/AtomicsObject.cpp (hunk 11 above)
Comment (akroshg) [inline lib/Runtime/Library/AtomicsObject.cpp]: Yes, the call to `max` will you give 0. I should have test for that. 

---
In reply to: [77405070](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77405070) [](ancestors = 77405070)

File: lib/Runtime/Library/AtomicsObject.cpp (hunk 12 above)
Comment (akroshg) [inline lib/Runtime/Library/AtomicsObject.cpp]: Sorry I didn't understand you meant? GetByteOffset is the value fetch.

---
In reply to: [77405200](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77405200) [](ancestors = 77405200)

File: lib/Runtime/Library/AtomicsObject.cpp (hunk 13 above)
Comment (akroshg) [inline lib/Runtime/Library/AtomicsObject.cpp]: Yes - good catch.

---
In reply to: [77406158](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77406158) [](ancestors = 77406158)

File: lib/Runtime/Library/Chakra.Runtime.Library.vcxproj (hunk 14 above)
Comment (akroshg) [inline lib/Runtime/Library/Chakra.Runtime.Library.vcxproj]: will do.

---
In reply to: [77406543](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77406543) [](ancestors = 77406543)

File: lib/Runtime/Library/JavascriptLibrary.cpp [hunk 16]
```diff
@@ -201,6 +201,21 @@ namespace Js
             DynamicType::New(scriptContext, TypeIds_Object, typedArrayPrototype, nullptr,
                 DeferredTypeHandler<InitializeCharArrayPrototype, DefaultDeferredTypeFilter, true>::GetDefaultInstance()));
 
+        if (scriptContext->GetConfig()->IsESSharedArrayBufferEnabled())
+        {
+            sharedArrayBufferPrototype = DynamicObject::New(recycler,
+                DynamicType::New(scriptContext, TypeIds_Object, objectPrototype, nullptr,
+                    DeferredTypeHandler<InitializeSharedArrayBufferPrototype, DefaultDeferredTypeFilter, true>::GetDefaultInstance()));
+
+            sharedArrayBufferType = DynamicType::New(scriptContext, TypeIds_SharedArrayBuffer, sharedArrayBufferPrototype, nullptr,
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/JavascriptLibrary.cpp]: >sharedArrayBufferType [](start = 12, length = 21)

Types are initialized in InitializeTypes(), please remove from here. Also move arrayBufferType Line#140 to InitializeTypes().

File: lib/Runtime/Library/JavascriptLibrary.cpp [hunk 17]
```diff
@@ -1630,6 +1673,66 @@ namespace Js
...
+        }
+
+        sharedArrayBufferPrototype->SetHasNoEnumerableProperties(true);
+    }
+
+    void  JavascriptLibrary::InitializeAtomicsObject(DynamicObject* atomicsObject, DeferredTypeHandlerBase * typeHandler, DeferredInitializeMode mode)
+    {
+        typeHandler->Convert(atomicsObject, mode, 12);
+
+        //ScriptContext* scriptContext = atomicsObject->GetScriptContext();
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/JavascriptLibrary.cpp]: >//ScriptContext* scriptContext = atomicsObject->GetScriptContext(); [](start = 8, length = 67)

nit: remove

File: lib/Runtime/Library/SharedArrayBuffer.cpp [hunk 18]
```diff
@@ -0,0 +1,577 @@
...
+
+        Assert(newBuffer);
+        Assert(newBuffer->GetByteLength() >= newbyteLength);
+
+        // Don't bother doing memcpy if we aren't copying any elements
+        if (newbyteLength > 0)
+        {
+            AssertMsg(currentBuffer->GetBuffer() != nullptr, "buffer must not be null when we copy from it");
+
+            js_memcpy_s(newBuffer->GetBuffer(), newbyteLength, currentBuffer->GetBuffer() + start, newbyteLength);
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/SharedArrayBuffer.cpp]: See ArrayBuffer::EntrySlice, we have bunch of IsDetached checks and I think those are applicable here as well. Can you make sure we have analyzed all detach scenarios.

File: lib/Runtime/Library/SharedArrayBuffer.cpp [hunk 19]
```diff
@@ -0,0 +1,577 @@
...
+
+    BYTE* SharedArrayBuffer::GetBuffer() const
+    {
+        return sharedContents != nullptr ? sharedContents->buffer : nullptr;
+    }
+
+    BOOL SharedArrayBuffer::GetDiagTypeString(StringBuilder<ArenaAllocator>* stringBuilder, ScriptContext* requestContext)
+    {
+        stringBuilder->AppendCppLiteral(_u("Object, (SharedArrayBuffer)"));
+        return TRUE;
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/SharedArrayBuffer.cpp]: I didn't see any debugger test, please add them under core\test\DebuggerCommon

File: lib/Runtime/Library/AtomicsObject.cpp (hunk 12 above)
Comment (agarwal-sandeep) [inline lib/Runtime/Library/AtomicsObject.cpp]: What I mean if ToNumber/ToInt32 calls above call into user code and user deletes the buffer, we should throw error and not do any of below operations.

---
In reply to: [77408403](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77408403) [](ancestors = 77408403,77405200)

File: lib/Parser/rterrors.h (hunk 2 above)
Comment (agarwal-sandeep) [inline lib/Parser/rterrors.h]: ok then please have A as capital in above error message :-)

---
In reply to: [77407079](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77407079) [](ancestors = 77407079,77394931)

File: lib/Runtime/Library/SharedArrayBuffer.cpp (hunk 19 above)
Comment (akroshg) [inline lib/Runtime/Library/SharedArrayBuffer.cpp]: Tools related tests will be added in the next sets of PRs. I haven't forgotten that :)

---
In reply to: [77417713](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77417713) [](ancestors = 77417713)

File: lib/Runtime/Library/SharedArrayBuffer.cpp (hunk 18 above)
Comment (akroshg) [inline lib/Runtime/Library/SharedArrayBuffer.cpp]: The SharedArrayBuffer does not get detached.

---
In reply to: [77416686](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77416686) [](ancestors = 77416686)

File: lib/Runtime/Library/JavascriptLibrary.cpp (hunk 16 above)
Comment (akroshg) [inline lib/Runtime/Library/JavascriptLibrary.cpp]: Not completely sure about the flow here. I have copied as the exact manner on how ArrayBufferType was created. If just moving does not work out then I'll leave this as is and investigate later.

---
In reply to: [77414757](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77414757) [](ancestors = 77414757)

File: lib/Runtime/Library/AtomicsObject.cpp (hunk 12 above)
Comment (akroshg) [inline lib/Runtime/Library/AtomicsObject.cpp]: I need to understand how user deletes the buffer. I have test (look for "Atomics.add buffer/length invalidation" in the atomics_test.js file), that does not delete that. Like to add the test case if you meant differently.

---
In reply to: [77417944](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77417944) [](ancestors = 77417944,77408403,77405200)

Comment (agarwal-sandeep) [inline lib/Runtime/Library/AtomicsObject.cpp]: What happens if you do ArrayBuffer.transfer(view, 10) in valueOf?

---
In reply to: [77425384](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77425384) [](ancestors = 77425384,77417944,77408403,77405200)

Comment (akroshg) [inline lib/Runtime/Library/AtomicsObject.cpp]: Arraybuffer.transfer can work on the ArrayBuffer only. It is a typeerror for other objects.

**Comment by agarwal-sandeep** [comment]
:shipit:

**Comment by akroshg** [comment]
Thanks @agarwal-sandeep for having a look.
@abchatra @Microsoft/chakra-es - would you also like to review?

**Comment by suwc** [comment]
How does perf compare?

File: lib/Runtime/Library/SharedArrayBuffer.cpp [hunk 20]
```diff
@@ -0,0 +1,577 @@
...
+                JavascriptError::ThrowRangeError(
+                    scriptContext, JSERR_ArrayLengthConstructIncorrect);
+            }
+            return byteCount;
+        }
+
+        return JavascriptConversion::ToUInt32(length, scriptContext);
+    }
+
+    Var SharedArrayBuffer::NewInstance(RecyclableObject* function, CallInfo callInfo, ...)
```
Comment (suwc) [inline lib/Runtime/Library/SharedArrayBuffer.cpp]: >NewInstance [](start = 27, length = 11)

This and several others are almost identical to their ArrayBuffer counterparts. Is it better to re-factor them into ArrayBufferBase?

File: lib/Runtime/Library/TypedArray.cpp [hunk 21]
```diff
@@ -450,7 +450,9 @@ namespace Js
                     }
                     else if (JavascriptConversion::ToObject(firstArgument, scriptContext, &jsArraySource))
                     {
-                        HRESULT hr = scriptContext->GetHostScriptContext()->ArrayBufferFromExternalObject(jsArraySource, &arrayBuffer);
+                        ArrayBuffer *temp = nullptr;
+                        HRESULT hr = scriptContext->GetHostScriptContext()->ArrayBufferFromExternalObject(jsArraySource, &temp);
+                        arrayBuffer = static_cast<ArrayBuffer *> (temp);
```
Comment (suwc) [inline lib/Runtime/Library/TypedArray.cpp]: >ArrayBuffer [](start = 50, length = 11)

ArrayBufferBase?

**Comment by akroshg** [comment]
I don't expect anything to change perf-wise (why do you there will be perf related issue), but I'll run just in case.

---
In reply to: [245406564](https://github.com/Microsoft/ChakraCore/pull/1533#issuecomment-245406564) [](ancestors = 245406564)

File: lib/Runtime/Library/SharedArrayBuffer.cpp (hunk 20 above)
Comment (akroshg) [inline lib/Runtime/Library/SharedArrayBuffer.cpp]: There are few subtle differences in these functions. I feel like it is best to keep them separate as the spec could evolve and it would be easier to make those changes. Majority of this functions are input validation. I could make them a common macro. 

---
In reply to: [77910245](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77910245) [](ancestors = 77910245)

File: lib/Runtime/Library/TypedArray.cpp (hunk 21 above)
Comment (akroshg) [inline lib/Runtime/Library/TypedArray.cpp]: Yes - will change that.

---
In reply to: [77910603](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77910603) [](ancestors = 77910603)

**Comment by akroshg** [comment]
I ran it and it is flat.

---
In reply to: [245458779](https://github.com/Microsoft/ChakraCore/pull/1533#issuecomment-245458779) [](ancestors = 245458779,245406564)

File: lib/Runtime/Library/SharedArrayBuffer.cpp (hunk 20 above)
Comment (ianwjhalliday) [inline lib/Runtime/Library/SharedArrayBuffer.cpp]: Would template usage help?

File: lib/Parser/rterrors.h [hunk 22]
```diff
@@ -357,3 +357,10 @@ RT_ERROR_MSG(JSERR_InvalidHint, 5658, "%s: invalid hint", "invalid hint", kjstTy
 
 RT_ERROR_MSG(JSERR_This_NeedNamespace, 5659, "%s: 'this' is not a Module Namespace object", "Module Namespace object expected", kjstTypeError, JSERR_This_NeedNamespace) // {Locked="\'this\'"}
 RT_ERROR_MSG(JSERR_This_NeedListIterator, 5660, "%s: 'this' is not a List Iterator object", "List Iterator expected", kjstTypeError, 0) 
+RT_ERROR_MSG(JSERR_NeedSharedArrayBufferObject, 5661, "%s is not a SharedArrayBuffer", "SharedArrayBuffer object expected", kjstTypeError, 0)
+
+RT_ERROR_MSG(JSERR_Function_LessArguments, 5662, "Function '%s' is called with less arguments", "Function called with less arguments", kjstRangeError, 0)
```
Comment (tcare) [inline lib/Parser/rterrors.h]: Is the intention here 'not enough arguments'? I would reword it. Perhaps we can use the existing RT_ERROR_MSG(JSERR_WinRTFunction_TooFewArguments, 5102, "%s: function called with too few arguments", "Function called with too few arguments", kjstError, 0)

File: lib/Runtime/Library/ArrayBuffer.h [hunk 23]
```diff
@@ -103,8 +128,12 @@ namespace Js
         static const uint32 MaxArrayBufferLength = 1 << 30;
 #endif
         virtual bool IsValidAsmJsBufferLength(uint length, bool forceCheck = false) { return false; }
-        virtual bool IsValidVirtualBufferLength(uint length) { return false; }
+        virtual bool IsArrayBuffer() override { return true; }
+        virtual bool IsSharedArrayBuffer() override { return false; }
+        virtual ArrayBuffer * GetAsArrayBuffer() override { return ArrayBuffer::FromVar((Var)this); }
```
Comment (tcare) [inline lib/Runtime/Library/ArrayBuffer.h]: >(Var) [](start = 88, length = 5)

nit: use a C++ style cast

Comment (tcare) [inline lib/Runtime/Library/ArrayBuffer.h]: Elsewhere too

---
In reply to: [78116315](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r78116315) [](ancestors = 78116315)

**Comment by akroshg** [comment]
@digitalinfinity can you look at the pal changes? thanks,

**Comment by digitalinfinity** [comment]
PAL changes LGTM
